docs/source/extension/plugin_system.rst

```diff
@@ -0,0 +1,179 @@
+.. Copyright (c) Jupyter Development Team.
+.. Distributed under the terms of the Modified BSD License.
+
+The JupyterLab Plugin System
+============================
+
+JupyterLab's plugin system is designed so that plugins can depend on and
+reuse features from one another. A key part of this approach is JupyterLab's
+**provider-consumer pattern**.
+
+The Provider-Consumer Pattern
+-----------------------------
+
+..
+    TODO add to glossary, provider-consumer, service objects, tokens
+
+In the provider-consumer pattern, one plugin "provides" an object (called a
+**service object**) to the system, and other plugins "consume" that object
+by using it in an extension to add extra features and customizations
+to JupyterLab (and an extension consists of one or more plugins).
+
+How Plugin Metadata Drives the Provider-Consumer Pattern
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Each plugin uses some properties (the "requires" and "optional" properties) to
+request features it wants which are provided by other plugins that have been
+loaded into JupyterLab. When your plugin requests features, the system sends
+them to your plugin's activate function if they're available.
+
+See the example plugin below, which is a "consumer" of the "IStatusBar":
+
+.. code::
+
+  const plugin: JupyterFrontEndPlugin<void> = {
+    id: 'shout_button_message:plugin',
+    description: 'An extension that adds a button to the right toolbar',
+    autoStart: true,
+    optional: [IStatusBar],
+    activate: (app: JupyterFrontEnd, statusBar: IStatusBar | null) => {
+      console.log('JupyterLab extension shout_button_message is activated!');
+
+      // Create a ShoutWidget and add it to the interface in the right sidebar
+      const shoutWidget: ShoutWidget = new ShoutWidget(statusBar);
+      shoutWidget.id = 'JupyterShoutWidget';  // Widgets need an id
+      app.shell.add(shoutWidget, 'right');
+    }
+  };
+
+Here, you can see the "optional" property, which is a list of optional
+features this plugin wants (with just a single item, IStatusBar). You
+can also see the "activate" property of the plugin, which is a callable
+(function) that the plugin system will call for you when your plugin is
+loaded.
+
+About the "activate" function
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+It's important to note that the arguments to your "activate" callable will depend on what
+things you request in your "optional" and "requires" plugin properties, so
+remember to add arguments for any service objects you request into your
+activate function's arguments.
+
+When JupyterLab calls your plugin's "activate" function, it will always
+pass an application as the first argument, then it will pass any "required"
+objects (in the order you specify them), then any "optional" objects (again,
+in the order you specify them).
+
+By returning an object from your activate function, you become a "provider"
+in Jupyter's provider-consumer pattern, and other plugins can use ("consume")
```

Comment:
Maybe use "JupyterLab" instead of "Jupyter"? So it's consistent with the rest of the docs in https://jupyterlab.readthedocs.io/en/latest/extension/extension_dev.html#plugins?

Reply:
This doc was moved to a new PR here, so I added your suggestion there.